[SYCL] vec abi unification and trivially copyable

cperkinsintel opened the pull request:
adapted from https://github.com/intel/llvm/pull/6816   

sycl::vec now uses std::array on host in all cases, but device access is via `ext_vector_type` so as to not lose performance.  Alignment is set the same as the size to a maximum of 64 bytes, which aligns with an upcoming spec change proposal. In the cases where the size is greater than 64 we use std::array on the device.  We no longer need to emit a pragma on Windows and we can use `alignas()` on all platforms.

cperkinsintel commented:
Hmm - am getting an internal error on Linux when it's trying to preprocess the builtins macros.  But I can't reproduce this error locally. The CI is using GCC 11.3.0 , but that works for me.  I also tested  GCC 9.4 and  Clang 17.0, both worked as well.

```
2023-05-17T17:29:36.6656716Z FAILED: tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_relational.cpp.o 
2023-05-17T17:29:36.6660930Z CCACHE_DIR=/__w//build_cache_default CCACHE_MAXSIZE=8G CCACHE_CPP2=yes CCACHE_HASHDIR=yes /usr/bin/ccache /usr/bin/g++ -DCL_TARGET_OPENCL_VERSION=300 -DENABLE_OPAQUE_POINTERS=1 -DSYCL2020_DISABLE_DEPRECATION_WARNINGS -DXPTI_ENABLE_INSTRUMENTATION -DXPTI_STATIC_LIBRARY -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D_LIBCPP_ENABLE_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__SYCL_INTERNAL_API -I/__w/llvm/llvm/build/tools/sycl/source -I/__w/llvm/llvm/src/sycl/source -I/__w/llvm/llvm/build/include -I/__w/llvm/llvm/src/llvm/include -I/__w/llvm/llvm/src/xpti/include -I/__w/llvm/llvm/src/sycl/include -I/__w/llvm/llvm/build/_deps/ocl-headers-src -I/__w/llvm/llvm/src/sycl-fusion/jit-compiler/include -I/__w/llvm/llvm/src/sycl-fusion/common/include -I/__w/llvm/llvm/build/_deps/vc-intrinsics-src/GenXIntrinsics/include -I/__w/llvm/llvm/build/_deps/vc-intrinsics-build/GenXIntrinsics/lib/GenXIntrinsics/../../include -I/__w/llvm/llvm/build/_deps/vc-intrinsics-build/GenXIntrinsics/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wextra -Wno-deprecated-declarations -Werror -O3 -DNDEBUG -UNDEBUG -fvisibility=hidden -fvisibility-inlines-hidden -Wno-psabi -std=c++17 -MD -MT tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_relational.cpp.o -MF tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_relational.cpp.o.d -o tools/sycl/source/CMakeFiles/sycl_object.dir/detail/builtins_relational.cpp.o -c /__w/llvm/llvm/src/sycl/source/detail/builtins_relational.cpp
2023-05-17T17:29:36.6664050Z In file included from /__w/llvm/llvm/src/sycl/source/detail/builtins_relational.cpp:12:
2023-05-17T17:29:36.6664817Z /__w/llvm/llvm/src/sycl/source/detail/builtins_relational.cpp: In function 'sycl::_V1::vec<int, 4> __host_std::sycl_host_SignBitSet(sycl::_V1::vec<float, 4>)':
2023-05-17T17:29:36.6665251Z /__w/llvm/llvm/src/sycl/source/detail/builtins_helper.hpp:24:3: error: unrecognizable insn:
2023-05-17T17:29:36.6665521Z    24 |   }
2023-05-17T17:29:36.6665695Z       |   ^
2023-05-17T17:29:36.6666233Z /__w/llvm/llvm/src/sycl/source/detail/builtins_helper.hpp:121:3: note: in expansion of macro '__MAKE_1V'
2023-05-17T17:29:36.6666589Z   121 |   __MAKE_1V(Fun, Call, 4, Ret, Arg1)                                           \
2023-05-17T17:29:36.6666813Z       |   ^~~~~~~~~
2023-05-17T17:29:36.6667239Z /__w/llvm/llvm/src/sycl/source/detail/builtins_relational.cpp:470:1: note: in expansion of macro 'MAKE_1V_FUNC'
2023-05-17T17:29:36.6667620Z   470 | MAKE_1V_FUNC(sycl_host_SignBitSet, __vSignBitSet, s::cl_int, s::cl_float)
2023-05-17T17:29:36.6667874Z       | ^~~~~~~~~~~~
2023-05-17T17:29:36.6668091Z (insn 10 7 11 2 (set (reg:V4SI 82 [ vect__17.61304 ])
2023-05-17T17:29:36.6668569Z         (lshiftrt:V4SI (subreg:V4SI (subreg:V4SF (reg/v:TI 88 [ x ]) 0) 0)
2023-05-17T17:29:36.6668941Z             (const_int 31 [0x1f]))) "/usr/include/c++/11/cmath":662:29 -1
2023-05-17T17:29:36.6669170Z      (nil))
2023-05-17T17:29:36.6669355Z during RTL pass: vregs
2023-05-17T17:29:36.6669677Z /__w/llvm/llvm/src/sycl/source/detail/builtins_helper.hpp:24:3: internal compiler error: in extract_insn, at recog.c:2770
2023-05-17T17:29:36.6669975Z    24 |   }
2023-05-17T17:29:36.6670138Z       |   ^
2023-05-17T17:29:36.6670541Z /__w/llvm/llvm/src/sycl/source/detail/builtins_helper.hpp:121:3: note: in expansion of macro '__MAKE_1V'
2023-05-17T17:29:36.6670878Z   121 |   __MAKE_1V(Fun, Call, 4, Ret, Arg1)                                           \
2023-05-17T17:29:36.6671095Z       |   ^~~~~~~~~
2023-05-17T17:29:36.6671514Z /__w/llvm/llvm/src/sycl/source/detail/builtins_relational.cpp:470:1: note: in expansion of macro 'MAKE_1V_FUNC'
2023-05-17T17:29:36.6671886Z   470 | MAKE_1V_FUNC(sycl_host_SignBitSet, __vSignBitSet, s::cl_int, s::cl_float)
2023-05-17T17:29:36.6672132Z       | ^~~~~~~~~~~~
2023-05-17T17:29:36.6672340Z 0xe3320d internal_error(char const*, ...)
2023-05-17T17:29:36.6672545Z 	???:0
2023-05-17T17:29:36.6672759Z 0xe29d35 fancy_abort(char const*, int, char const*)
2023-05-17T17:29:36.6672963Z 	???:0
2023-05-17T17:29:36.6673208Z 0xb017cc _fatal_insn(char const*, rtx_def const*, char const*, int, char const*)
2023-05-17T17:29:36.6673462Z 	???:0
2023-05-17T17:29:36.6673693Z 0xb017ee _fatal_insn_not_found(rtx_def const*, char const*, int, char const*)
2023-05-17T17:29:36.6673927Z 	???:0
2023-05-17T17:29:36.6674124Z Please submit a full bug report,
2023-05-17T17:29:36.6674369Z with preprocessed source if appropriate.
2023-05-17T17:29:36.6674657Z Please include the complete backtrace with any bug report.
2023-05-17T17:29:36.6675072Z See <file:///usr/share/doc/gcc-11/README.Bugs> for instructions.
```

cperkinsintel commented on sycl/include/sycl/detail/generic_type_traits.hpp:
```diff
@@ -378,8 +378,27 @@ template <typename T> class TryToGetPointerVecT {
+      using DT = typename DecoratedType<elementType
+                                        __attribute__((ext_vector_type(16))),
+                                        T::address_space>::type *;
+      return DT();
```
@romanovvlad Here I'm fixing the work group copying problem. It's a bit ungainly. Also it is introducing a warning when SYCL is built ( `warning: ignoring attributes applied to dependent type ‘elementType’ without an associated declaration [-Wattributes]` ).

Wanted to get your opinion before I start tackling formatting and any more test case updates.

romanovvlad replied:
Personally I would prefer trying to split 128 byte vec to two 64 byte vectors and call builtins twice.
But if we go with casting to 128 byte vector. Not sure I fully understand why it doesn't work out of the box, these helpers use `vec::vector_t`. I guess for > 64 byte types `vec::vector_t` is `std::array` which doesn't seem to be correct. If it so, wouldn't changing `vec::vector_t` for > 64 byte types to corresponding `ext_vector_type` fix the issue with async copy builtin?

cperkinsintel replied:
AFAICT the `vector_t` operator is not used by the helpers. They use `.get()`.

Having `vector_t` return the vector extension seems correct, but I'm not sure it matters - I don't think it has any callers.

AlexeySachkov commented on sycl/include/sycl/detail/vector_traits.hpp:
```diff
@@ -17,12 +17,15 @@ inline namespace _V1 {
-                         std::integral_constant<int, sizeof(T) * N>> {};
+    : std::conditional_t<
+          N == 3,
+          std::integral_constant<size_t, (std::min)(sizeof(T) * 4, (size_t)64)>,
```
```suggestion
          std::integral_constant<size_t, (std::min)(sizeof(T) * 4, 64u)>,
```

?

AlexeySachkov commented on sycl/include/sycl/types.hpp:
```diff
@@ -580,6 +553,41 @@ template <typename Type, int NumElements> class vec {
+      typename detail::VecStorage<DataT, NumElements>::VectorDataType;
+
+  VectorDataType &getAsVector() {
+    return *reinterpret_cast<VectorDataType *>(m_Data.data());
```
We should leave a comment here that breaking strict aliasing rules is a UB. Doing this could lead to weird issues depending on how optimizations work

rolandschulz replied:
If the size is below 64 bytes, strict aliasing is the only UB problem. Is my understanding correct that all our supported compilers (MSVC, gcc, clang) don't produce incorrect results for this (because while UB it is too common in real code)? 

Above 64 bytes this is also UB because the alignment is wrong. Can this produce incorrect results?

cperkinsintel replied:
@rolandschulz  the max alignment is 64 (as was agreed in the JIRA discussion) and that's what I've implemented here, and in that case we switch to std::array instead of ext_vector_type. So, no incorrect results on that front.

rolandschulz replied:
The max alignment of 64 is for the storage. The storage is  using `DataType` which is `ext_vector_type` up to 64 bytes,  and array with 64 byte alignment otherwise. But the return type is `VectorDataType`. And `VectorDataType` is always `ext_vector_type` (if supported by compiler). Is all this correct? 
If so the problem is that you are returning a reference with a type with >64 byte alignment which refers to storage with 64 byte alignment. In the JIRA we didn't discuss what we do with the return type of getAsVector.
I don't see how we can implement getAsVector with the spec as-is.

It would be possible if we were to change the spec to say:
- `getAsVector` is only available up to 64 bytes, or
- return by value rather than reference, or
- return by value above 64 bytes

@gmlueck Do you think one of the changes would be acceptable?

rolandschulz replied:
Sorry I confused `getAsVector` with `operator vector_t()`. Only the later is in the spec. And it returns it by value. Therefore I agree this isn't a spec concern. 
We still need to make sure that the internal usage gives correct results for >64bytes.

rolandschulz replied:
I only see one usage of getAsVector (impl of binop). If we replace that with bitcast we don't have either UB problem.

AlexeySachkov commented on sycl/include/sycl/types.hpp:
```diff
@@ -580,6 +553,41 @@ template <typename Type, int NumElements> class vec {
+
+  static constexpr size_t AdjustedNum = (NumElements == 3) ? 4 : NumElements;
+  static constexpr size_t Sz = sizeof(DataT) * AdjustedNum;
+  static constexpr bool IsSizeGreaterThanAlign = (Sz > 64);
```
```suggestion
  static constexpr bool IsSizeGreaterThanMaxAlign = (Sz > 64);
```

? Should we add a comment about what this magic number 64 means?

cperkinsintel replied:
I have added a constexpr var to replace 64 and added a comment about it at its declaration in vector_traits.hpp

AlexeySachkov commented on sycl/include/sycl/types.hpp:
```diff
@@ -907,8 +903,23 @@ template <typename Type, int NumElements> class vec {
+    if constexpr (!IsUsingArray) {
+      return m_Data;
+    } else {
+      auto ptr = reinterpret_cast<const VectorDataType *>((&m_Data)->data());
+      return *ptr;
```
We should probably use `sycl::bit_cast` here to avoid breaking strict aliasing rules

AlexeySachkov commented on sycl/include/sycl/types.hpp:
```diff
@@ -2165,18 +2174,37 @@ template <typename T, int N, typename V> struct VecStorage {
+#define __SYCL_DEFINE_VECSTORAGE_IMPL(type, cl_type, num)                      \
+  template <> struct VecStorageImpl<type, num> {                               \
+    using DataType = std::array<type, (num == 3) ? 4 : num>;                   \
+    using VectorDataType = std::array<type, (num == 3) ? 4 : num>;             \
```
```suggestion
    using VectorDataType = DataType;             \
```

?

AlexeySachkov commented on sycl/include/sycl/types.hpp:
```diff
@@ -2208,18 +2238,26 @@ struct VecStorage<bool, N, typename std::enable_if_t<isValidVectorSize(N)>> {
 struct VecStorage<T, 1, typename std::enable_if_t<is_sigeninteger<T>::value>> {
   using DataType = select_apply_cl_t<T, std::int8_t, std::int16_t, std::int32_t,
                                      std::int64_t>;
+  using VectorDataType = select_apply_cl_t<T, std::int8_t, std::int16_t,
+                                           std::int32_t, std::int64_t>;
```
```suggestion
  using VectorDataType = DataType;
```

?

cperkinsintel replied:
done, and got the others as well. thanks.

AlexeySachkov commented on sycl/source/CMakeLists.txt:
```diff
@@ -92,6 +92,7 @@ function(add_sycl_rt_library LIB_NAME LIB_OBJ_NAME)
   else()
     target_compile_options(${LIB_OBJ_NAME} PUBLIC
                            -fvisibility=hidden -fvisibility-inlines-hidden)
+    target_compile_options(${LIB_OBJ_NAME} PUBLIC -Wno-psabi)
```
In-sources comment with a motivation/description would be helpful here, I think

cperkinsintel replied:
done

AlexeySachkov reviewed (commented):
We probably need a note in a commit message, that strictly speaking, this breaks current SYCL 2020 spec, but that we are expecting it to change accordingly. Link to the spec change would also be useful

romanovvlad commented on sycl/include/sycl/types.hpp:
```diff
@@ -2165,18 +2174,37 @@ template <typename T, int N, typename V> struct VecStorage {
+#else // __SYCL_DEVICE_ONLY__
 // When ext_vector_type is not available, we rely on cl_* types from CL/cl.h
 // to represent vec storage.
+#ifdef __HAS_EXT_VECTOR_TYPE__
```
I believe we should always use `std::array` on the host, otherwise we may break when translation units compiled using different compilers.

cperkinsintel replied:
Ok, I've made the requested change to std::array only on host  and I removed the now irrelevant `__NO_EXT_VECTOR_TYPE_ON_HOST__` macro.

romanovvlad commented on sycl/include/sycl/types.hpp:
```diff
@@ -907,8 +903,23 @@ template <typename Type, int NumElements> class vec {
+  }
+
+  operator vector_t() const {
+    if constexpr (!IsUsingArray) {
```
So, we never convert to `ext_vector_size` when the size if bigger than 64? Right?

cperkinsintel commented on sycl/include/sycl/types.hpp:
```diff
@@ -907,8 +903,23 @@ template <typename Type, int NumElements> class vec {
+      m_Data = openclVector;
+    } else {
+      detail::loop<NumElements>(
+          [&, this](size_t i) { m_Data[i] = openclVector[i]; });
```
good catch.

steffenlarsen commented on sycl/source/detail/builtins_relational.cpp:
```diff
@@ -458,9 +458,15 @@ __SYCL_EXPORT rel_res_t sycl_host_SignBitSet(s::cl_float x) __NOEXC {
 __SYCL_EXPORT s::cl_int __vSignBitSet(s::cl_float x) __NOEXC {
   return -static_cast<s::cl_int>(std::signbit(x));
 }
+#endif
```
Discussed offline. This is in library source code and is likely used but is missing test coverage, so we would be breaking ABI by conditionally excluding it here.

cperkinsintel replied:
@steffenlarsen  - the ABI symbol table is updated to reflect that, and it's not the only ABI change in this PR. Users will definitely have to recompile. Attempting to get a dispensation.

I checked this out yesterday, and this exact case ( vec<float> -> int)  IS under test. I think it passes because there are other overloads and the compiler is implicitly converting. The overloads are all identical.  Ultimately, I think this OK. Let me know if you disagree - I can make sycl::signbitset a template and move it into the headers. Just need to find a place to sequester a solitary math built-in.  I think the better solution is to, one day, move all the math built-ins to templates and lose these macro expansions. But that's out of scope for this.

steffenlarsen replied:
You might be right. It could be that we have an implicit promotion of `cl_float` to `cl_double` and then a demotion from `cl_long` to `cl_int`. I might worry that this could cause warnings on some compilers, so you might need to be ready for promotion/demotion warnings in post-commit.

Looking at it, I don't see why we would even export the `__v*` functions. Seems to me like they are simply used inside other exported functions. Had the ABI-break window not been closed we should've removed `__SYCL_EXPORT` from these. Next time, maybe!

> the ABI symbol table is updated to reflect that, and it's not the only ABI change in this PR. Users will definitely have to recompile. Attempting to get a dispensation.

Assuming you are right and we can safely mask this out, wouldn't the symbols check when we compile the library with clang (e.g. in post-commit) as the symbol would suddenly be back?

steffenlarsen commented on sycl/source/detail/builtins_relational.cpp:
```diff
@@ -458,9 +458,15 @@ __SYCL_EXPORT rel_res_t sycl_host_SignBitSet(s::cl_float x) __NOEXC {
 __SYCL_EXPORT rel_res_t sycl_host_SignBitSet(s::cl_double x) __NOEXC {
   return std::signbit(x);
 }
+#ifndef __GNUC__
+// GCC 11.3.0 (and friends) has an internal compiler error when this function is
+// passed to the macro declaration below. ( MAKE_1V ... 4 ... ) Fortunately, we
+// don't actually need it when compiling the SYCL library itself.
+// TODO: switch to templates instead of these overload expansions.
 __SYCL_EXPORT s::cl_int __vSignBitSet(s::cl_float x) __NOEXC {
   return -static_cast<s::cl_int>(std::signbit(x));
 }
+#endif
```
I tested it locally and I believe we can do this without breaking ABI, by simply doing the redirection that we would do anyway.

```suggestion
__SYCL_EXPORT s::cl_int __vSignBitSet(s::cl_float x) __NOEXC {
  return static_cast<s::cl_int>(__vSignBitSet(static_cast<s::cl_double>(x)));
}
```

Alternatively, if we want to preserve the old version when building with a non-GCC compiler, we can do something like:

```suggestion
__SYCL_EXPORT s::cl_int __vSignBitSet(s::cl_float x) __NOEXC {
#ifdef __GNUC__
  return static_cast<s::cl_int>(__vSignBitSet(static_cast<s::cl_double>(x)));
#else
  return -static_cast<s::cl_int>(std::signbit(x));
#endif
}
```

Feel free to leave a comment about the work-around and the TODO to move them.

romanovvlad commented on sycl/include/sycl/types.hpp:
```diff
@@ -33,14 +26,7 @@
-#define __SYCL_USE_EXT_VECTOR_TYPE__
-#endif
-#elif defined(__SYCL_DEVICE_ONLY__)
+#if !defined(__HAS_EXT_VECTOR_TYPE__) && defined(__SYCL_DEVICE_ONLY__)
```
Could you please clarify why we need to check for `__HAS_EXT_VECTOR_TYPE__`? What unique cases it covers?

cperkinsintel replied:
I think I'm misunderstanding your question. 

This full block is
```
#if !defined(__HAS_EXT_VECTOR_TYPE__) && defined(__SYCL_DEVICE_ONLY__)
// This is a soft error. We expect the device compiler to have ext_vector_type
// support, but that should not be a hard requirement.
#error "SYCL device compiler is built without ext_vector_type support"
#endif // __HAS_EXT_VECTOR_TYPE__
```

And we are erroring if the device compiler does not have ext_vector_type support, because we need it. We don't check at all on the host compiler where we are no longer using it.  

Let me know if I'm missing your intent.

romanovvlad replied:
But we also have it checked a couple of times in the file. And depending on that either use `ext_vector_type` or not.

cperkinsintel replied:
I think I see what you are after.  Yes, now that we are always using using `ext_vector_type` on device and never using it on host, the inclusion of `__HAS_EXT_VECTOR_TYPE` in some of the guards is now redundant. I have removed them. Pretty much just used for that error output now.

romanovvlad commented on sycl/include/sycl/types.hpp:
```diff
@@ -56,7 +42,9 @@
+// When ext_vector_type is not available, we rely on cl_* types from CL/cl.h
+// to represent vec storage.
+#if !defined(__HAS_EXT_VECTOR_TYPE__) || !defined(__SYCL_DEVICE_ONLY__)
 #include <sycl/detail/cl.h>
```
Where do we use cl_* types?

cperkinsintel replied:
We need both a DataType and a VectorDataType.  The std::array on the host, now that we are not using ext_vector_type, needs a VectorDataType, so we use the cl_ types for that access. We need something that the math built-ins support already.  Does that make sense? I didn't test this specific inclusion though - it was there before. Maybe it's not necessary.

cperkinsintel commented:
Ping to reviewers,

I believe I've addressed all the concerns, either here or in conversations offline. They are holding the ABI-breaking window open for this PR, so I'd like keep the process moving along.  Let me know if you need anything else.

romanovvlad reviewed (approved):
LGTM

steffenlarsen commented:
Post-commit found that the redirected call will cause a recursion as the double overload is defined later than the call to it. To avoid this https://github.com/intel/llvm/pull/10652 moves said overload definition.
